src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs

```diff
@@ -2497,5 +2996,260 @@ namespace System
+        var comp = CreateCompilation(source1);
+        comp.VerifyDiagnostics(
+            // (3,16): error CS9303: File type 'Outer.C' must be defined in a top level type; 'Outer.C' is a nested type.
+            //     file class C { } // 1
+            Diagnostic(ErrorCode.ERR_FileTypeNested, "C").WithArguments("Outer.C").WithLocation(3, 16));
+        var sourceMember = comp.GetMember<NamedTypeSymbol>("Outer.C");
+        Assert.Equal("<>F0__C", sourceMember.MetadataName);
+
+        var sourceType = comp.GetTypeByMetadataName("Outer.<>F0__C");
+        // Note: strictly speaking, it would be reasonable to return the (invalid) nested file type symbol here.
+        // However, since we don't actually support nested file types, we don't think we need the API to do the additional lookup
+        // when the requested type is nested, and so we end up giving a null here.
+        Assert.Null(sourceType);
+    }
+
+    [Fact]
+    public void GetTypeByMetadataName_04()
+    {
+        var source1 = """
+            file class C { }
+            """;
+
+        var source2 = """
+            class C { }
+            """;
+
+        // from source
+        var comp = CreateCompilation(new[] { source1, source2 });
+        comp.VerifyDiagnostics();
+        var sourceMember = comp.GetMembers("C")[0];
+        Assert.Equal("<>F0__C", sourceMember.MetadataName);
+
+        var sourceType = comp.GetTypeByMetadataName("<>F0__C");
+        Assert.Equal(sourceMember, sourceType);
+
+        // from metadata
+        var comp2 = CreateCompilation("", references: new[] { comp.EmitToImageReference() });
+        comp2.VerifyDiagnostics();
+
+        var metadataMember = comp2.GetMember<NamedTypeSymbol>("<>F0__C");
+        Assert.Equal("<>F0__C", metadataMember.MetadataName);
+
+        var metadataType = comp2.GetTypeByMetadataName("<>F0__C");
+        Assert.Equal(metadataMember, metadataType);
+    }
+
+    [CombinatorialData]
+    [Theory]
+    public void GetTypeByMetadataName_05(bool firstIsMetadataReference, bool secondIsMetadataReference)
+    {
+        var source1 = """
+            file class C { }
+            """;
+
+        // Create two references containing identically-named file types
+        var ref1 = CreateCompilation(source1, assemblyName: "ref1");
+        var ref2 = CreateCompilation(source1, assemblyName: "ref2");
+
+        var comp = CreateCompilation("", references: new[]
+        {
+            firstIsMetadataReference ? ref1.ToMetadataReference() : ref1.EmitToImageReference(),
+            secondIsMetadataReference ? ref2.ToMetadataReference() : ref2.EmitToImageReference()
+        });
+        comp.VerifyDiagnostics();
+
+        var sourceType = comp.GetTypeByMetadataName("<>F0__C");
+        Assert.Null(sourceType);
+
+        var types = comp.GetTypesByMetadataName("<>F0__C");
+        Assert.Equal(2, types.Length);
+        Assert.Equal(firstIsMetadataReference ? "C@<tree 0>" : "<>F0__C", types[0].ToTestDisplayString());
```

Comment:
>"<>F0__C" [](http://example.com/codeflow?start=63&length=9)

This is not blocking, but it feels strange that display is different

Reply:
Added test plan item